Sources/tart/OCI/Layerizer/DiskV2.swift

```diff
@@ -198,14 +209,22 @@ class DiskV2: Disk {
 
     var offset = offset
 
-    for chunk in data.chunks(ofCount: holeGranularityBytes) {
+    for chunk in data.subdataChunks(ofCount: holeGranularityBytes) {
       // If the local layer cache is used, only write chunks that differ
       // since the base disk can contain anything at any position
       if let rdisk = rdisk {
-        // F_PUNCHHOLE requires the holes to be aligned to file system block boundaries
-        let isHoleAligned = (offset % fsBlockSize) == 0 && (UInt64(chunk.count) % fsBlockSize) == 0
+        try rdisk.seek(toOffset: offset)
+        let actualContentsOnDisk = try rdisk.read(upToCount: chunk.count)
 
-        if isHoleAligned && chunk == zeroChunk {
+        if actualContentsOnDisk == chunk {
```

Comment:
This is still a place for optimization. One though was to not try to optimize wholes on the sub-layer level, only try to deduplicate layers. And in this case whole punch the whole uncompressed layer and then sparse write into it.